pylabrobot/plate_reading/molecular_devices_backend.py

```diff
@@ -0,0 +1,1154 @@
+  """A collection of MolecularDevicesDataAbsorbance objects from multiple reads."""
+
+  reads: List["MolecularDevicesDataAbsorbance"]
+  all_absorbance_wavelengths: List[int]
+
+
+@dataclass
+class MolecularDevicesDataCollectionFluorescence(MolecularDevicesDataCollection):
+  """A collection of MolecularDevicesDataFluorescence objects from multiple reads."""
+
+  reads: List["MolecularDevicesDataFluorescence"]
+  all_excitation_wavelengths: List[int]
+  all_emission_wavelengths: List[int]
+
+
+@dataclass
+class MolecularDevicesDataCollectionLuminescence(MolecularDevicesDataCollection):
+  """A collection of MolecularDevicesDataLuminescence objects from multiple reads."""
+
+  reads: List["MolecularDevicesDataLuminescence"]
+  all_emission_wavelengths: List[int]
+
+
+class MolecularDevicesBackend(PlateReaderBackend, metaclass=ABCMeta):
+  """Backend for Molecular Devices plate readers."""
+
+  def __init__(self, port: str, res_term_char: bytes = b">") -> None:
+    self.port = port
+    self.io = Serial(self.port, baudrate=9600, timeout=0.2)
+    self.res_term_char = res_term_char
+
+  async def setup(self) -> None:
+    await self.io.setup()
+    await self.send_command("!")
+
+  async def stop(self) -> None:
+    await self.io.stop()
+
+  def serialize(self) -> dict:
+    return {**super().serialize(), "port": self.port}
+
+  async def send_command(
+    self, command: str, timeout: int = 60, num_res_fields=None
+  ) -> MolecularDevicesResponse:
+    """Send a command and receive the response, automatically determining the number of
+    response fields.
+    """
+    base_command = command.split(" ")[0]
+    if num_res_fields is None:
+      num_res_fields = COMMAND_TERMINATORS.get(base_command, 1)
+    else:
+      num_res_fields = max(1, num_res_fields)
+
+    await self.io.write(command.encode() + b"\r")
+    raw_response = b""
+    timeout_time = time.time() + timeout
+    while True:
+      raw_response += await self.io.readline()
+      await asyncio.sleep(0.001)
+      if time.time() > timeout_time:
+        raise TimeoutError(f"Timeout waiting for response to command: {command}")
+      if raw_response.count(self.res_term_char) >= num_res_fields:
+        break
+    logger.debug("[plate reader] Command: %s, Response: %s", command, raw_response)
+    response = raw_response.decode("utf-8").strip().split(self.res_term_char.decode())
+    response = [r.strip() for r in response if r.strip() != ""]
+    self._parse_basic_errors(response, command)
+    return response
+
+  async def _send_commands(self, commands: List[Union[Optional[str], Tuple[str, int]]]) -> None:
+    """Send a sequence of commands to the plate reader."""
+    for command_info in commands:
+      if not command_info:
+        continue
+      if isinstance(command_info, tuple):
+        command, num_res_fields = command_info
+        await self.send_command(command, num_res_fields=num_res_fields)
+      else:
+        await self.send_command(command_info)
```

Comment:
what do you think about calling send_command on every line as opposed to making a list then calling using a helper function

is there any logic somewhere that manipulates the list? I can't immediately find any instances

it would really simplify the code to send the commands directly